File: 백준/Gold/2250. 트리의 높이와 너비/트리의 높이와 너비.java
import java.util.*;
import java.io.*;

public class Main {
    static int idx = 0;
    static Map<Integer, int[]> level = new HashMap<>();
    static int[] left, right, par;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        par = new int[N + 1];
        left = new int[N + 1];
        right = new int[N + 1];
        for (int i = 1; i <= N; i++) {
            par[i] = i;
        }

        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());
            int l = Integer.parseInt(st.nextToken());
            int r = Integer.parseInt(st.nextToken());
            if (l != -1) {
                left[n] = l;
                par[l] = n;
            }
            if (r != -1) {
                right[n] = r;
                par[r] = n;
            }
        }

        for (int i = 1; i <= N; i++) {
            if (par[i] == i) {
                inorder(i, 1);
                break;
            }
        }

        int ansLevel = 10001, ansWidth = 0;
        for (Map.Entry<Integer, int[]> entry : level.entrySet()) {
            int levelKey = entry.getKey();
            int width = entry.getValue()[1] - entry.getValue()[0] + 1;
            if (width > ansWidth || (width == ansWidth && levelKey < ansLevel)) {
                ansLevel = levelKey;
                ansWidth = width;
            }
        }

        System.out.println(ansLevel + " " + ansWidth);
    }

    static void inorder(int n, int l) {
        if (n == 0) return;
        inorder(left[n], l + 1);
        idx++;
        if (level.containsKey(l)) {
            level.get(l)[1] = idx;
        } else {
            level.put(l, new int[]{idx, idx});
        }
        inorder(right[n], l + 1);
    }
}

File: 백준/Gold/2250. 트리의 높이와 너비/트리의 높이와 너비.js
function inorder(n, l) {
    if (!n) return;
    inorder(left[n], l+1);
    idx++;
    if (level[l]) {
        level[l][1] = idx;
    } else {
        level[l] = [idx, idx];
    }
    inorder(right[n], l+1);
}

const [[N], ...ipt] = require('fs').readFileSync(0).toString().trim().split('\n').map(x=>x.split(' ').map(Number));
const par = Array.from({ length: N+1 }, (_, i) => i);
const left = Array(N+1).fill(0);
const right = Array(N+1).fill(0);

for (const [n, l, r] of ipt) {
  if (l !== -1) {
      left[n] = l;
      par[l] = n;
  }
  if (r !== -1) {
      right[n] = r;
      par[r] = n;
  }
}

let idx = 0;
const level = {};
for (let i=1; i<=N; i++) {
  if (i === par[i]) {
      inorder(i, 1);
      break;
  }
}

let ansLevel = 10001;
let ansWidth = 0;
for (const k of Object.keys(level).map(Number).sort((a,b)=>a-b)) {
    const width = level[k][1] - level[k][0] + 1;
    if (ansWidth < width) {
        ansLevel = k;
        ansWidth = width;
    }
}

console.log(ansLevel, ansWidth);
